File: Java/65. Valid Number.java
class Solution {
    public boolean isNumber(String s) {
        boolean hasDigit = false, hasEe = false, hasDot = false;
        char c;
        
        for (int i=0; i<s.length(); i++) {
            c = s.charAt(i);
            
            if (Character.isDigit(c)) {
                hasDigit = true;
            }
            
            else if (c == '+' || c == '-') {
                // signs can only present at 0-th position
                // and right after 'e' or 'E'
                if (i > 0 && s.charAt(i-1) != 'e' && s.charAt(i-1) != 'E') {
                    return false;
                }
            }
            
            else if (c == 'e' || c == 'E') {
                // exponent must be surrounded with digits
                if (hasEe || !hasDigit) {
                    return false;
                }
                hasEe = true;
                hasDigit = false;
            }
            
            else if (c == '.') {
                // can only have one dot and must be before the exponent if applicable
                if (hasDot || hasEe) {
                    return false;
                }
                hasDot = true;
            }
            
            else {
                // all other characters are invalid
                return false;
            }
        }
        return hasDigit;
    }
}

File: Python3/65. Valid Number.py
class Solution:
    def isNumber(self, s: str) -> bool:
        hasDot = hasEx = hasDigit = False
        
        for i, c in enumerate(s):
            if c.isdigit():
                hasDigit = True
                
            elif c in '+-':
                if i > 0 and s[i-1] != 'e' and s[i-1] != 'E':
                    return False
            
            elif c in 'eE':
                if hasEx or not hasDigit:
                    return False
                hasEx = True
                hasDigit = False
                
            elif c == '.':
                if hasEx or hasDot:
                    return False
                hasDot = True
            
            else:
                return False
            
        return hasDigit